### PR TITLE
sourceHashes json

### PR DESCRIPTION
**Pull request by delphiactual**
Proof of concept for issues raised in #1303...

This only adds in `Speeder` && `Queen's Guard` Gear... More will be added if this implementation is acceptable...

### CODE REVIEW

File: app/scripts/services/dimStoreService.factory.js [hunk 1]
```diff
@@ -1593,5 +1602,22 @@
...
+
+    function loadJSON(file) {
+      return $http.get(file)
+               .then(function(json) {
+                 return json.data.itemHash;
+               });
+    }
+
+    function getMissingSourceHashes(itemHash, missingSources) {
+      if (missingSources.$$state.value) {
```
Comment (bhollis) [inline app/scripts/services/dimStoreService.factory.js]: This isn't the right way to use a promise. You want to resolve it using `then` (or add it as an argument to a `$q.all`).

File: app/scripts/sources.json [hunk 2]
```diff
@@ -0,0 +1,157 @@
+{
+    "itemHash":{
+       "488703545":{"sourceHashes":[1662396737]},
```
Comment (bhollis) [inline app/scripts/sources.json]: I'm not super comfortable with the idea of it being our responsibility to  augment or correct Bungie's data.

That said, this whole file could be much simpler, right?

```json
{
  "488703543": [1662396737],
  "4063984059": [1662396737],
}
```

File: app/scripts/services/dimStoreService.factory.js (hunk 1 above)
Comment (delphiactual) [inline app/scripts/services/dimStoreService.factory.js]: can you give me an example usage, ive tried changing `loadJSON` to
```
    function loadJSON(file) {
      return $http.get(file)
               .then(function(json) {
                 return json.data
                 .then(function(data) {
                   return data;
                 });
               });
    }
```
and the promise is still not resolving...

File: app/scripts/sources.json (hunk 2 above)
Comment (delphiactual) [inline app/scripts/sources.json]: It was more about supplementing easily known bad/missing data...

File: app/scripts/services/dimStoreService.factory.js [hunk 3]
```diff
@@ -1593,5 +1602,19 @@
...
     // code above is from https://github.com/DestinyTrialsReport
+
+    function loadJSON(file) {
+      return $http.get(file)
+               .then(function(json) {
+                 return json.data;
+               });
+    }
+
+    function getMissingSourceHashes(itemHash, missingSources) {
```
Comment (bhollis) [inline app/scripts/services/dimStoreService.factory.js]: I think `missingSources` will be a promise here, so you need to use `then` to asynchronously access its data. Best to pass `missingSources` into one of the `$q.all` resolvers like the one that takes `dimBucketService` and `dimDefinitions` up top.

**Comment by delphiactual** [comment]
@bhollis awesome thanks for that.

File: app/scripts/services/dimStoreService.factory.js [hunk 4]
```diff
@@ -336,15 +336,16 @@
...
         dimItemInfoService(activePlatform),
         dimBungieService.getStores(activePlatform)])
-        .then(function([defs, buckets, newItems, itemInfoService, rawStores]) {
+        .then(function([defs, buckets, missingSources, newItems, itemInfoService, rawStores]) {
           console.timeEnd('Load stores (Bungie API)');
           if (activePlatform !== dimPlatformService.getActive()) {
             throw new Error("Active platform mismatch");
           }
-
+          dimMissingSources = missingSources;
```
Comment (bhollis) [inline app/scripts/services/dimStoreService.factory.js]: Reassigning `dimMissingSources` from a promise to its resolved value isn't a great idea. Just pass along the resolved value to where it needs to go.

File: app/scripts/services/dimStoreService.factory.js [hunk 5]
```diff
@@ -706,6 +710,11 @@
 
       itemDef.sourceHashes = itemDef.sourceHashes || [];
 
+      var missingSource = getMissingSourceHashes(itemDef.hash, missingSources);
```
Comment (bhollis) [inline app/scripts/services/dimStoreService.factory.js]: This line could be:

```js
const missingSource = missingSources[itemDef.hash] || [];
```

No extra function needed.

File: app/scripts/services/dimStoreService.factory.js [hunk 6]
```diff
@@ -1593,5 +1603,19 @@
       return ret;
     }
     // code above is from https://github.com/DestinyTrialsReport
+
+    function loadJSON(file) {
```
Comment (bhollis) [inline app/scripts/services/dimStoreService.factory.js]: This probably doesn't need its own function either.

**Comment by bhollis** [comment]
Nice, thanks for iterating on this!

**Comment by delphiactual** [comment]
@bhollis no problem thanks for the pointers

**Pull request merged**